### ISSUE TITLE
Add Typescript support to Node SDK

### ISSUE THREAD

**Issue by WillGarman** [amadeus4dev/amadeus-node#212]
## Description
Add typescript support to Node SDK. 

## Reason
Currently the Node SDK has 0 type support and make developer experience tough when building an app with typescript and trying to use the SDK

it would be nice to be able to see what an expected response would be in the code editor. This also causes responses to be typed as `any` which isnt quite right since we can see the types in the documentation

![image](https://github.com/amadeus4dev/amadeus-node/assets/73144742/f50155e4-f236-447e-90c9-1b8ada120047)

## Solution
We currently use `npm i amadeus --save` to install the SDK. Adding something like `npm i --save-dev @types/amadeus` would be amazing

**Comment by marenga14** [comment]
I see there's no response for this, but it's really needed

**Comment by slk333** [comment]
second this, would help a lot.

or at least provide declaration files so that we get autocomplete in VS Code

**Comment by slk333** [comment]
- clone this repo
- run `tsc --declaration --allowJs --emitDeclarationOnly --outDir ./@types src/**/*.js`
now you have a directory called `@types` with 2 items in it

- now go to your personal project, open the `node_modules`, look for the `@types` directory
- drop the 2 items from you generated before into this directory

now it works automatically

**Comment by marenga14** [comment]
Okay let me check, on my side i just added the .d.ts file, but i still get
the responses with no types defaulted to type any

**Comment by bedros-p** [comment]
Yeah, the developer experience is not great with the Node SDK... I'd rather use the REST API than this typeless library

It's odd though, all this JSDoc in the files you'd think there'd be *some* level of decent typing?
Zero autocomplete in my code editor is absolutely a deal breaker for me.

**Comment by bedros-p** [comment]
Just checked the closed issues - https://github.com/amadeus4dev/amadeus-node/issues/73#issuecomment-1233963555 


> This repository is not based on Typescript.
> For that reason, we are going to close this issue.

⁉️

**Comment by slk333** [comment]
they use babel before publishing the package on npm, and they don't include the src directory.

but if you clone this repository, you have access to the src directory, and you can the following command on it, to generate declaration files, from the JS source code:

`tsc --declaration --allowJs --emitDeclarationOnly --outDir ./@types src/**/*.js`

You then need to do some shenanigans to import this to your project. I managed it myself, but it's a bit cumbersome

**Comment by slk333** [comment]
[drop those in src.zip](https://github.com/user-attachments/files/16502911/drop.those.in.src.zip)

I zipped the declaration files so you just have to unzip and drop the content in your `src` directory.

Let me know if it works

**Comment by slk333** [comment]
After unzipping and dropping the files, your `src` directory should like this

<img width="215" alt="Screenshot 2024-08-05 at 23 29 42" src="https://github.com/user-attachments/assets/b9ce6356-83c6-41c5-88c2-9afbf7fdd9b8">

**Comment by slk333** [comment]
I'm sorry for the notifications, the indications I posted before seem to have an issue.

I won't post again before i'm 100% sure I understand how to reproduce.

The only way i have it to work is to drop it in the node_modules/@types directory but this gets deleted everytime we npm install something, so it's not worth making a guide.

**Comment by bedros-p** [comment]
There really should be a fork of this by now that has this issue solved, seems like this is a problem going for 2 years now

**Comment by darseen** [comment]
I have created [amadeus-ts](https://www.npmjs.com/package/amadeus-ts) package. I rewrote the entire codebase in TypeScript, and so far I have added types for the most commonly used endpoint which is flight offers search and flight offers pricing.
The package is stable but still under development, and I'm trying to add new types for the remaining endpoints each day.

I Have tested the library and it works perfectly fine apart from the incomplete types for the remaining endpoints that I'm currently working on. Please test the library and if you find any issues please submit them in the GitHub repository and any PRs are welcome as well.

**Comment by darseen** [comment]
The package [amadeus-ts](https://github.com/darseen/amadeus-ts) is complete. It includes types for all endpoints now.

**Comment by bedros-p** [comment]
> The package [amadeus-ts](https://github.com/darseen/amadeus-ts) is complete. It includes types for all endpoints now.

You are everything open source stands for
Thank you 🫡

**Comment by darseen** [comment]
> > The package [amadeus-ts](https://github.com/darseen/amadeus-ts) is complete. It includes types for all endpoints now.
> 
> You are everything open source stands for Thank you 🫡

You're welcome, I'm glad I was able to help.

**Comment by m-partida** [comment]
Catching up on this thread and seeing there is a need for a TypeScript SDK provided by Amadeus. Happy to work with the Amadeus team to build an officially support supported SDK, that would help developers with faster time to integration. This would also remove the burden/risk of requiring the community to maintain an SDK that is constantly evolving.

For reference, we recently worked with Mistral to build their TS SDK: [https://github.com/mistralai/client-ts](https://github.com/mistralai/client-ts)

**Comment by olle-johannesson** [comment]
Perhaps the most low-impact way would be to autogenerate the types from JSDoc on the top-level API of the package and publish it to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped). Add a git hook to check the JSDoc, and this should be a sufficient first step, and then we can work from there.

Changing the programming language of a project is a significant decision, so it’s no surprise if @darseen’s wonderful contribution is met with insufficient enthusiasm.
